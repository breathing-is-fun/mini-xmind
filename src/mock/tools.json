[
  {
    "key": "minus",
    "value": "line"
  },
  {
    "key": "border",
    "value": "border",
    "style": {
      "width": 100,
      "height": 80,
      "background": "#F96",
      "borderRadius": "10px",
      "border": "1px solid #aaa"
    }
  }
]