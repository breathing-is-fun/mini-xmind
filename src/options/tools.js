// key: for type of Icon
// value: id
// style: class of tool's wrapper
export const tools = [
  {
    key: 'minus',
    value: 'line',
  },
  {
    key: 'border',
    value: 'block',
    style: {
      width: 100,
      height: 80,
      background: '#F96',
      borderRadius: '10px',
      border: '1px solid #aaa',
    },
  },
  {
    key: 'tag',
    value: 'input',
    style: {
      width: 100,
      height: 32,
    },
  },
];
