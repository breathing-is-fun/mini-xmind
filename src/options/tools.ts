export const tools = [
  {
    key: 'minus',
    value: 'line',
  },
  {
    key: 'border',
    value: 'block',
  },
  {
    key: 'tag',
    value: 'input',
    editable: true,
  },
];
