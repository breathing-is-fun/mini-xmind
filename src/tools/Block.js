import React, { Component } from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';

import './assets/Block.css';

export default class Block extends Component {
  static propTypes = {
    className: PropTypes.string,
  };

  static defaultProps = {
    className: '',
  };

  render() {
    const { className, ...rest } = this.props;
    return <div className={classNames('Block', className)} {...rest} />;
  }
}
