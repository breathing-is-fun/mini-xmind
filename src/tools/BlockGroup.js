import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import Draggable from 'react-draggable';
import { noop } from '../utils/commonUtil';
import { generateKey, stopPropagation } from '../utils/LineUtil';
import omit from 'omit.js';
import './assets/BlockGroup.css';

// one Line is mapping to two Block
// to record it here
let mapping = {};

let blockDOM = {};

const addBlockDom = (lineData, blockDOM) => {
  for (let key in lineData) {
    const { fromKey, toKey } = lineData[key];

    for (let blockKey in blockDOM) {
      const value = blockDOM[blockKey];

      if (fromKey == blockKey) {
        lineData[key].from = value;
      }

      if (toKey == blockKey) {
        lineData[key].to = value;
      }
    }
  }
  return lineData;
};

export default class Block extends Component {
  static propTypes = {
    className: PropTypes.string,
    data: PropTypes.object,
    lineData: PropTypes.object,
    onChange: PropTypes.func,
  };

  static defaultProps = {
    className: '',
    data: {},
    lineData: {},
    onChange: noop,
  };

  constructor(props) {
    super(props);

    this.state = {
      data: {},
      lineData: {},
    };

    // for Line, two point create a line
    this.checkBlockClickList = {};
  }

  static getDerivedStateFromProps(nextProps, nextState) {
    const { data, onChange } = nextProps;

    mapping = Object.assign({}, mapping, nextProps.lineData);
    if (
      Object.keys(nextProps.lineData).length !=
      Object.keys(nextState.lineData).length
    ) {
      // hack: when lineData is loaded, blockDOM is not mounted
      setTimeout(() => {
        const lineData = addBlockDom(nextProps.lineData, blockDOM);
        onChange && onChange(data, lineData);
      }, 0);
    }

    return {
      data,
      lineData: nextProps.lineData,
    };
  }

  componentDidMount = () => {
    const { props } = this;
    const { lineData } = props;

    addBlockDom(lineData, blockDOM);
  };

  handleStop = ({ x, y }, blockKey) => {
    const { data, onChange } = this.props;

    data[blockKey] = Object.assign({}, data[blockKey], { x, y });

    onChange && onChange(data);
  };

  // when Block clicked twice, generate a Line
  // and clear checkBlockClickList
  handleBlockClick = blockKey => {
    let { checkBlockClickList } = this;
    let { onChange } = this.props;
    const { lineData, data } = this.state;
    const lineKey = generateKey('line');

    checkBlockClickList[blockKey] = { current: blockDOM[blockKey] };

    // to know which Block is starting point
    if (!('time' in checkBlockClickList[blockKey])) {
      checkBlockClickList[blockKey].time = new Date().getTime();
    }

    if (Object.keys(checkBlockClickList).length == 2) {
      if (!this.shouldPaintLine(checkBlockClickList, lineData)) {
        this.checkBlockClickList = {};
        return;
      }

      const { result, fromKey, toKey } = this.generateLineData(
        lineData,
        lineKey,
      );

      onChange && onChange(data, result);

      this.checkBlockClickList = {};
      // record mapping for arrow
      mapping[lineKey] = {
        fromKey,
        toKey,
      };
    }
  };

  generateLineData = (lineData, lineKey) => {
    const { checkBlockClickList } = this;
    let fromNode, toNode, fromKey, toKey;
    const keys = Object.keys(checkBlockClickList);

    if (checkBlockClickList[keys[0]] > checkBlockClickList[keys[1]]) {
      fromKey = keys[1];
      toKey = keys[0];
    } else {
      fromKey = keys[0];
      toKey = keys[1];
    }

    fromNode = checkBlockClickList[fromKey].current;
    toNode = checkBlockClickList[toKey].current;

    const common = {
      fromKey,
      toKey,
    };
    lineData[lineKey] = {
      ...common,
      from: fromNode,
      to: toNode,
    };

    return {
      result: lineData,
      ...common,
    };
  };

  saveBlock = (ref, blockKey) => {
    blockDOM[blockKey] = ReactDOM.findDOMNode(ref);
  };

  shouldPaintLine = (checkBlockClickList, linesProps) => {
    if (!Object.keys(linesProps).length) {
      return true;
    }

    const blocks = Object.keys(checkBlockClickList).toString();
    for (let key of Object.keys(mapping)) {
      let fromFlag = false,
        toFlag = false;
      const { fromKey, toKey } = mapping[key];

      if (blocks.includes(fromKey)) {
        fromFlag = true;
      }

      if (blocks.includes(toKey)) {
        toFlag = true;
      }

      if (fromFlag && toFlag) {
        return false;
      }
    }

    return true;
  };

  handleDragStart = e => {
    stopPropagation(e);
  };

  render() {
    const { className, onChange, ...rest } = this.props;
    const { data } = this.state;

    DataCollector.set('BlockGroup', data);
    return Object.keys(data).map(blockKey => {
      const { x, y, style } = data[blockKey];
      return (
        <Draggable
          onStop={(e, item) => this.handleStop(item, blockKey)}
          key={blockKey}
          position={{ x, y }}
          onDrag={e => onChange && onChange(data)}
          onStart={this.handleDragStart}
        >
          <div
            className={classNames('BlockGroup', 'animate-appear', className)}
            onClick={e => this.handleBlockClick(blockKey)}
            ref={ref => this.saveBlock(ref, blockKey)}
            style={style}
            {...omit(rest, ['lineData'])}
          />
        </Draggable>
      );
    });
  }
}
