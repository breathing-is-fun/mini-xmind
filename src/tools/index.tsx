import Toolbar from './Toolbar';

export default Toolbar;
export { default as TagGroup } from './TagGroup';
export { default as BlockGroup } from './BlockGroup';
