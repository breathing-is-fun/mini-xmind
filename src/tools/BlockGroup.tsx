import React, { Component } from 'react';
import classNames from 'classnames';
import Draggable from 'react-draggable';
import { generateKey, stopPropagation } from '../utils/LineUtil';
import './css/BlockGroup.css';
import { ContextMenuProps } from '../canvas/core';

// one Line is mapping to two Block
// to record it here
let mapping: any = {};
// to save refs
const blockDOM: any = {};
const keysLength = (obj: object) => Object.keys(obj).length;

const addBlockDom = (lineData: any, targetBlockDOM: any) => {
  for (const key of Object.keys(lineData || {})) {
    const { fromKey, toKey } = lineData[key];
    for (const blockKey of Object.keys(targetBlockDOM)) {
      const value = targetBlockDOM[blockKey];

      if (fromKey === blockKey) {
        lineData[key].from = value;
      }

      if (toKey === blockKey) {
        lineData[key].to = value;
      }
    }
  }
  return lineData;
};

export type BlockProps = { [blockKey: string]: { x: number; y: number } };
export interface BlockGroupProps {
  className?: string;
  data?: BlockProps;
  lineData?: any;
  onChange?: (data: any, lineData?: any) => void;
  offset?: { x: number; y: number };
  onContextMenu?: (item: ContextMenuProps) => void;
  renderLine?: (lineData: any) => void;
}
export interface BlockGroupState {
  data?: any;
  lineData?: any;
}

export default class BlockGroup extends Component<
  BlockGroupProps,
  BlockGroupState
> {
  static getDerivedStateFromProps(nextProps: BlockGroupProps) {
    mapping = Object.assign({}, mapping, nextProps.lineData);
    return {
      data: nextProps.data,
      lineData: nextProps.lineData,
    };
  }

  state: BlockGroupState = {
    data: {},
    lineData: {},
  };

  private checkBlockClickList: any = {};

  componentDidUpdate = (prevProps: BlockGroupProps) => {
    const { lineData, onChange, data } = this.props;
    if (keysLength(lineData) === 0) {
      return;
    }
    const firstLine: any = Object.values(lineData)[0];
    const hasNewLine =
      keysLength(lineData) !== keysLength(prevProps.lineData || {});
    if (!(firstLine && firstLine.from)) {
      if (hasNewLine || keysLength(lineData) !== 0) {
        if (onChange) {
          onChange(data, addBlockDom(lineData, blockDOM));
        }
      }
    }
    // it's a hack
    // when dragging canvas, LineData can't update
    // add this to force update
    // so Line will be repainted instantly
    if (this.props.offset && prevProps.offset) {
      if (
        this.props.offset.x !== prevProps.offset.x ||
        this.props.offset.y !== prevProps.offset.y
      ) {
        if (onChange) {
          onChange(data, addBlockDom(lineData, blockDOM));
        }
      }
    }
  };

  handleDrag = ({ x, y }: { x: number; y: number }, blockKey: string) => {
    const { data, onChange, lineData } = this.props;
    if (data) {
      data[blockKey] = Object.assign({}, data[blockKey], { x, y });
    }
    if (onChange) {
      onChange(data, addBlockDom(lineData, blockDOM));
    }
  };

  // when Block clicked twice, generate a Line
  // and clear checkBlockClickList
  handleBlockClick = (blockKey: string) => {
    const { checkBlockClickList } = this;
    const { onChange } = this.props;
    const { lineData, data } = this.state;
    const lineKey = generateKey('line');

    checkBlockClickList[blockKey] = { current: blockDOM[blockKey] };

    // to know which Block is starting point
    if (!('time' in checkBlockClickList[blockKey])) {
      checkBlockClickList[blockKey].time = new Date().getTime();
    }

    if (keysLength(checkBlockClickList) === 2) {
      if (!this.shouldPaintLine(checkBlockClickList, lineData)) {
        this.checkBlockClickList = {};
        return;
      }

      const { result, fromKey, toKey } = this.generateLineData(
        lineData,
        lineKey,
      );

      if (onChange) {
        onChange(data, result);
      }

      this.checkBlockClickList = {};
      // record mapping for arrow
      mapping[lineKey] = {
        fromKey,
        toKey,
      };
    }
  };

  generateLineData = (lineData: any, lineKey: string) => {
    const { checkBlockClickList } = this;
    let fromNode;
    let toNode;
    let fromKey;
    let toKey;
    const keys = Object.keys(checkBlockClickList);

    if (checkBlockClickList[keys[0]] > checkBlockClickList[keys[1]]) {
      fromKey = keys[1];
      toKey = keys[0];
    } else {
      fromKey = keys[0];
      toKey = keys[1];
    }

    fromNode = checkBlockClickList[fromKey].current;
    toNode = checkBlockClickList[toKey].current;

    const common = {
      fromKey,
      toKey,
    };
    lineData[lineKey] = {
      ...common,
      from: fromNode,
      to: toNode,
    };

    return {
      result: lineData,
      ...common,
    };
  };

  saveBlock = (ref: HTMLDivElement | null, blockKey: string) => {
    if (ref) {
      blockDOM[blockKey] = ref.getBoundingClientRect();
    }
  };

  shouldPaintLine = (checkBlockClickList: any, linesProps: any) => {
    if (!keysLength(linesProps)) {
      return true;
    }

    const blocks = Object.keys(checkBlockClickList).toString();
    for (const key of Object.keys(mapping)) {
      let fromFlag = false;
      let toFlag = false;
      const { fromKey, toKey } = mapping[key];

      if (blocks.includes(fromKey)) {
        fromFlag = true;
      }

      if (blocks.includes(toKey)) {
        toFlag = true;
      }

      if (fromFlag && toFlag) {
        return false;
      }
    }

    return true;
  };

  handleDragStart = (e: any) => {
    stopPropagation(e);
  };

  render() {
    const {
      className: parentClassName,
      onChange,
      onContextMenu,
      renderLine,
      lineData,
      ...rest
    } = this.props;
    const { data } = this.state;

    return (
      <>
        {Object.keys(data).map(blockKey => {
          const { x, y, className: blockClassName } = data[blockKey];
          return (
            <React.Fragment key={blockKey}>
              <Draggable
                position={{ x, y }}
                onDrag={(_, item) => this.handleDrag(item, blockKey)}
                onStart={this.handleDragStart}
              >
                <div
                  className={classNames(
                    'block-group',
                    'animate-appear',
                    parentClassName,
                    blockClassName,
                  )}
                  onClick={_ => this.handleBlockClick(blockKey)}
                  ref={ref => this.saveBlock(ref, blockKey)}
                  onContextMenu={(e: any) => {
                    if (onContextMenu) {
                      onContextMenu({
                        event: e,
                        key: blockKey,
                        group: 'BlockGroup',
                      });
                    }
                  }}
                  {...rest}
                />
              </Draggable>
            </React.Fragment>
          );
        })}
        {renderLine && renderLine(addBlockDom(lineData, blockDOM))}
      </>
    );
  }
}
