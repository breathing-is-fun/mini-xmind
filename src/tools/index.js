import Toolbar from './Toolbar';
import BlockGroup from './BlockGroup';

export default Toolbar;
export { BlockGroup };
