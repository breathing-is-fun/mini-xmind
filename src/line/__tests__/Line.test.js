import React from 'react';
import ReactDOM from 'react-dom';
import { render, mount, shallow } from 'enzyme';
import { SteppedLineTo } from '..';
import { Block } from '../../tools';

class Demo extends React.Component {
  constructor() {
    super();

    this.state = {
      steps: null,
    };
  }

  componentDidMount = () => {
    let steps = (
      <SteppedLineTo
        from={ReactDOM.findDOMNode(this.blockA)}
        to={ReactDOM.findDOMNode(this.blockB)}
        fromAnchor="bottom"
        toAnchor="top"
      />
    );

    this.setState({ steps });
  };

  render() {
    return (
      <React.Fragment>
        <Block
          style={{
            left: 90,
            top: 50,
            color: '#00f',
          }}
          ref={ref => (this.blockA = ref)}
          key="blockA"
        >
          A
        </Block>
        <Block
          style={{
            left: 20,
            top: 150,
            color: '#00f',
          }}
          ref={ref => (this.blockB = ref)}
          key="blockB"
        >
          B
        </Block>
        <Block
          style={{
            left: 100,
            top: 100,
            color: '#00f',
          }}
          ref={ref => (this.blockB = ref)}
          key="blockC"
        >
          B
        </Block>
        {this.state.steps}
      </React.Fragment>
    );
  }
}

describe('Line', () => {
  it('Block and SteppedLine render correctly', () => {
    const wrapper = mount(<Demo />);

    expect(wrapper).toMatchSnapshot();
  });
});
