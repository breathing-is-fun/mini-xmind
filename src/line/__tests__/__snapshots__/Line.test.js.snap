// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Line SteppedLine render correctly 1`] = `
<Demo>
  <Block
    className=""
    key="blockA"
    style={
      Object {
        "color": "#00f",
        "left": 90,
        "top": 50,
      }
    }
  >
    <div
      className="Block"
      style={
        Object {
          "color": "#00f",
          "left": 90,
          "top": 50,
        }
      }
    >
      A
    </div>
  </Block>
  <Block
    className=""
    key="blockB"
    style={
      Object {
        "color": "#00f",
        "left": 20,
        "top": 150,
      }
    }
  >
    <div
      className="Block"
      style={
        Object {
          "color": "#00f",
          "left": 20,
          "top": 150,
        }
      }
    >
      B
    </div>
  </Block>
  <SteppedLineTo
    from={
      <div
        class="Block"
        style="left: 90px; top: 50px; color: rgb(0, 0, 255);"
      >
        A
      </div>
    }
    fromAnchor="bottom"
    to={
      <div
        class="Block"
        style="left: 20px; top: 150px; color: rgb(0, 0, 255);"
      >
        B
      </div>
    }
    toAnchor="top"
  >
    <SteppedLine
      from={
        <div
          class="Block"
          style="left: 90px; top: 50px; color: rgb(0, 0, 255);"
        >
          A
        </div>
      }
      orientation="v"
      to={
        <div
          class="Block"
          style="left: 20px; top: 150px; color: rgb(0, 0, 255);"
        >
          B
        </div>
      }
      x0={0}
      x1={0}
      y0={0}
      y1={0}
    >
      <Line
        className=""
        from={
          <div
            class="Block"
            style="left: 90px; top: 50px; color: rgb(0, 0, 255);"
          >
            A
          </div>
        }
        orientation="v"
        style={Object {}}
        to={
          <div
            class="Block"
            style="left: 20px; top: 150px; color: rgb(0, 0, 255);"
          >
            B
          </div>
        }
        x0={0}
        x1={0}
        y0={0}
        y1={0}
      >
        <div
          className="react-lineto-placeholder"
        >
          <div
            className=""
            from={
              <div
                class="Block"
                style="left: 90px; top: 50px; color: rgb(0, 0, 255);"
              >
                A
              </div>
            }
            orientation="v"
            style={
              Object {
                "borderTopColor": "#ddd",
                "borderTopStyle": "solid",
                "borderTopWidth": 1,
                "left": "0px",
                "position": "absolute",
                "top": "0px",
                "transform": "rotate(0deg)",
                "transformOrigin": "0 0",
                "width": "0px",
                "zIndex": "1",
              }
            }
            to={
              <div
                class="Block"
                style="left: 20px; top: 150px; color: rgb(0, 0, 255);"
              >
                B
              </div>
            }
          />
        </div>
      </Line>
    </SteppedLine>
  </SteppedLineTo>
</Demo>
`;
